text/0000-pack-phase-debug-script.md

```diff
@@ -0,0 +1,114 @@
+- What use cases does it support?
+    - Example scenarios: 
+      - Trying to validate filesystem ownership/permissions/layout within a build container
+      - Validating cnb toml content pre-post phase execution to ensure changes are as expected
+      - Dumping checksum hashes of builder content to validate binaries being tested are the right ones
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The pack cli would allow specification of a debug script, that should be driven instead of the appropriate lifecycle binary for each phase. 
+
+The script will be passed the name/path and arguments of the lifecycle binary that would have been driven. 
+
+The script is able to run abitrary commands and dump information to stdout/stderr, these commands can include invoking the original lifecycle binary with its original arguments. 
+
+A simple debug script could look like.. 
+
+```bash
+      #!/bin/bash
+      echo "Layer dir Content Before"
+      ls -lar /layers      
+      LIFECYCLE=$1
+      shift
+      $LIFECYCLE "$@"
+      echo "Layer dir Content After"
+      ls -lar /layers  
+```
+
+This dumps the content of the /layers dir, invokes the orignal command and then dumps it again.. 
+
+From a technical perspective, the debug script will be added to the ephemeral builder image at /cnb/lifecycle/debug, and used as the command when invoking the phase container. 
+
+# Migration
+[migration]: #migration
+
+This is a new feature, no migration issues are anticipated.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Adds clutter to an already complex set of commandline options for pack
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  - custom debug buildpacks .. don't allow visibilty over changes by non-buildpack phases. 
+  - custom injected lifecycle binaries with extra debug added .. a pain to build/rebuild each time you need to verify soemthing else
+  - live snooping into the container .. difficult to ensure what you are viewing as the build is executing concurrently
+- Why is this proposal the best?
+  - it's the simplest way to allow developers to gain visibility into the build process
+- What is the impact of not doing this?
+  - harder to understand builds, harder to debug permission issues... etc
+
+# Prior Art
+[prior-art]: #prior-art
+
+Suggesting here, as I implemented this in the snowdrop platform implementation https://github.com/snowdrop/java-buildpack-client 
```

Comment:
I searched for "debug" on that page and it's mostly logging. Is there a better url/reference to read? Any docs for the feature?

Reply:
The snowdrop platform is a Java library, and the debug script is supplied as part of the platform config.. Probably easiest to see an example.. https://github.com/snowdrop/java-buildpack-client/blob/main/samples/hello-quarkus/pack.java

In my case I'm accepting a string holding the script, mostly to avoid issues caring about reading a file from disk to pass in.. it's an implementation detail that's relatively unimportant, and pack could easily implement it whichever way is easiest. 

The idea was useful enough I thought it worth suggesting for pack too.